src/ring_theory/witt_vector/truncated.lean

```diff
@@ -240,3 +238,221 @@ instance : comm_ring (truncated_witt_vector p n R) :=
+begin
+  rw [submodule.eq_bot_iff],
+  intros x hx,
+  ext,
+  simp only [witt_vector.mem_ker_truncate, ideal.mem_infi, witt_vector.zero_coeff] at hx ⊢,
+  exact hx _ _ (nat.lt_succ_self _)
+end
+
+end truncated_witt_vector
+
+namespace witt_vector
+open truncated_witt_vector (hiding truncate coeff)
+
+section lift
+
+variable [comm_ring R]
+variables {S : Type*} [comm_ring S]
+variable (f : Π k : ℕ, S →+* truncated_witt_vector p k R)
+variable f_compat : ∀ (k₁ k₂ : ℕ) (hk : k₁ ≤ k₂),
+           (truncated_witt_vector.truncate hk).comp (f k₂) = f k₁
+variables {p R}
+variable (n)
+
+/--
+Given a family `fₖ : S → truncated_witt_vector p k R` and `s : S`, we produce a Witt vector by
+defining the `k`th entry to be the final entry of `fₖ s`.
+-/
+def lift_fun (s : S) : 𝕎 R :=
+witt_vector.mk p $ λ k, truncated_witt_vector.coeff (fin.last k) (f (k+1) s)
+
+variables {f}
+include f_compat
+
+@[simp] lemma truncate_lift_fun (s : S) :
+  witt_vector.truncate n (lift_fun f s) = f n s :=
+begin
+  ext i,
+  simp only [lift_fun, truncated_witt_vector.coeff_mk, witt_vector.truncate_mk],
+  rw [← f_compat (i+1) n i.is_lt, ring_hom.comp_apply, truncated_witt_vector.coeff_truncate],
+  -- this is a bit unfortunate
+  congr' with _,
+  simp only [fin.coe_last, fin.coe_cast_le],
+end
+
+variable (f)
+
+/--
+Given compatible ring homs from `S` into `truncated_witt_vector n` for each `n`, we can lift these
+to a ring hom `S → 𝕎 R`.
+
+`lift` defines the universal property of `𝕎 R` as the inverse limit of `truncated_witt_vector n`.
+-/
+def lift : S →+* 𝕎 R :=
+by refine_struct { to_fun := lift_fun f };
+   { intros,
+     rw [← sub_eq_zero, ← ideal.mem_bot, ← infi_ker_truncate, ideal.mem_infi],
+     simp [ring_hom.mem_ker, f_compat] }
+
+variable {f}
+
+@[simp] lemma truncate_lift (s : S) :
+  witt_vector.truncate n (lift _ f_compat s) = f n s :=
+truncate_lift_fun _ f_compat s
+
+@[simp] lemma truncate_comp_lift :
+  (witt_vector.truncate n).comp (lift _ f_compat) = f n :=
+by { ext1, rw [ring_hom.comp_apply, truncate_lift] }
+
+/-- The uniqueness part of the universal property of `𝕎 R`. -/
+lemma lift_unique (g : S →+* 𝕎 R) (g_compat : ∀ k, (witt_vector.truncate k).comp g = f k) :
+  lift _ f_compat = g :=
+begin
+  ext1 x,
+  rw [← sub_eq_zero, ← ideal.mem_bot, ← infi_ker_truncate, ideal.mem_infi],
+  intro i,
+  simp only [ring_hom.mem_ker, g_compat, ←ring_hom.comp_apply,
+    truncate_comp_lift, ring_hom.map_sub, sub_self],
+end
+
+omit f_compat
+include hp
+
+/-- The universal property of `𝕎 R` as projective limit of truncated Witt vector rings. -/
+@[simps] def lift_equiv : (S →+* 𝕎 R) ≃ {f : Π k, S →+* truncated_witt_vector p k R //
+  ∀ k₁ k₂ (hk : k₁ ≤ k₂), (truncated_witt_vector.truncate hk).comp (f k₂) = f k₁} :=
+{ to_fun := λ g, ⟨λ k, (truncate k).comp g,
+    by { intros _ _ h, simp only [←ring_hom.comp_assoc, truncate_comp_witt_vector_truncate] }⟩,
+  inv_fun := λ f, lift f.1 f.2,
```

Comment:
It feels to me like the direction of this equiv should be flipped so that `lift_equiv.to_fun` is roughly `lift`

Reply:
hmm, that makes sense. I'll fix it.